v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/ConvertChangeStream.java

```diff
@@ -0,0 +1,291 @@
+            .equals(ByteString.copyFromUtf8(hbaseQualifierInput))) {
+          Metrics.counter("HbaseRepl", "mutations_filtered_from_hbase").inc();
+          return true;
+        }
+      }
+      Metrics.counter("HbaseRepl", "mutations_replicated_from_bigtable").inc();
+      return false;
+    }
+
+    static byte[] convertUtf8String(String utf8) {
+      return utf8.getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Appends origin information to row mutation.
+     *
+     * @param hbaseMutations row mutation to append origin info to
+     * @param cbtQualifierInput origin info string denoting mutation is from bigtable
+     * @throws IOException
+     */
+    private void appendSourceTagToMutations(RowMutations hbaseMutations, String cbtQualifierInput)
+        throws IOException {
+      byte[] lastEntryCf = getLastCellColFamily(hbaseMutations);
+      Delete hiddenDelete = new Delete(hbaseMutations.getRow(), 0L);
+      hiddenDelete.addColumns(lastEntryCf, convertUtf8String(cbtQualifierInput));
+      hbaseMutations.add(hiddenDelete);
+    }
+
+    /**
+     * Gets column family from last mutation of row mutations.
+     *
+     * @param hbaseMutations rowMutations object
+     * @return columnFamily byte array
+     * @throws IOException
+     */
+    private byte[] getLastCellColFamily(RowMutations hbaseMutations) throws IOException {
+      // het last mutation to get last column family from
+      Mutation lastMutation =
+          hbaseMutations.getMutations().get(hbaseMutations.getMutations().size() - 1);
+      CellScanner scanner = lastMutation.cellScanner();
+      // get to last cell
+      Cell cell = scanner.current();
+      while (scanner.advance()) {
+        cell = scanner.current();
+      }
+      return CellUtil.cloneFamily(cell);
+    }
+
+    /**
+     * Converts Cdc ChangeStreamMutation to HBase RowMutations object.
+     *
+     * @param mutation changeStreamMutation
+     * @return Hbase RowMutations object
+     */
+    private static RowMutations convertToRowMutations(ChangeStreamMutation mutation)
+        throws Exception {
+      byte[] hbaseRowKey = mutation.getRowKey().toByteArray();
+      RowMutations hbaseMutations = new RowMutations(hbaseRowKey);
+
+      for (Entry entry : mutation.getEntries()) {
+        if (entry instanceof SetCell) {
+
+          hbaseMutations.add(convertSetCell(hbaseRowKey, entry));
+
+        } else if (entry instanceof DeleteCells) {
+
+          hbaseMutations.add(convertDeleteCells(hbaseRowKey, entry));
+
+        } else if (entry instanceof DeleteFamily) {
+
+          hbaseMutations.add(convertDeleteFamily(hbaseRowKey, entry));
+        }
+      }
+
+      if (hbaseMutations.getMutations().size() == 0) {
+        throw new Exception(
+            "hbase mutations object null, should not be possible because all mutations are supported");
+      }
+      return hbaseMutations;
+    }
+
+    private static Put convertSetCell(byte[] hbaseRowKey, Entry entry) {
```

Comment:
I think it's worth unit testing each of these (even though they should also be covered by the ConvertChangeStreamTest)

Reply:
Added explicit unit tests in separate file to test ConvertChangeStreamMutationFn.convertToRowMutations function